Lib/_pyio.py

```diff
@@ -1794,6 +1794,21 @@ def isatty(self):
         self._checkClosed()
         return os.isatty(self._fd)
 
+    def _isatty_open_only(self):
+        """Checks whether the file is a TTY using an open-only optimization.
+
+        TTYs are always character devices. If the interpreter knows a file is
+        not a character device when it would call ``isatty``, can skip that
+        call. Inside ``open()``  there is a fresh stat result that contains that
+        information. Use the stat result to skip a system call. Outside of that
+        context TOCTOU issues (the fd could be arbitrarily modified by
+        surrounding code).
+        """
+        if (self._stat_atopen is not None
+            and not stat.S_ISCHR(self._stat_atopen.st_mode)):
+            return True
```

Comment:
I think this should be:

```suggestion
            return False
```

Reply:
Yep, `Py_RETURN_FALSE;` I got right in the C but not the pyio. Making a new PR to update....

Reply:
https://github.com/python/cpython/pull/125089